### ISSUE TITLE
CSpell: Add more words to the dictionary to avoid nagging in PRs

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#6255]
Another iteration after #6047.

Most PRs I've been rebasing recently a tripping in the following files/lines/words:

core/includes/bootstrap.inc:1668:58 Unknown word (pageviews)
core/includes/diff.inc:1053:18 Unknown word (xbeg)
core/includes/diff.inc:1054:18 Unknown word (xlen)
core/includes/diff.inc:1055:18 Unknown word (ybeg)
core/includes/diff.inc:1056:18 Unknown word (ylen)
core/includes/diff.inc:1061:29 Unknown word (xbeg)
core/includes/diff.inc:1061:36 Unknown word (xlen)
core/includes/diff.inc:1061:43 Unknown word (ybeg)
core/includes/diff.inc:1061:50 Unknown word (ylen)
core/includes/diff.inc:1062:43 Unknown word (xbeg)
core/includes/diff.inc:1062:50 Unknown word (xlen)
core/includes/diff.inc:1062:57 Unknown word (ybeg)
core/includes/diff.inc:1062:64 Unknown word (ylen)
core/includes/diff.inc:1098:14 Unknown word (xbeg)
core/includes/diff.inc:1099:14 Unknown word (xlen)
core/includes/diff.inc:1100:14 Unknown word (ybeg)
core/includes/diff.inc:1101:14 Unknown word (ylen)
core/includes/diff.inc:1105:35 Unknown word (xbeg)
core/includes/diff.inc:1105:42 Unknown word (xlen)
core/includes/diff.inc:1105:49 Unknown word (ybeg)
core/includes/diff.inc:1105:56 Unknown word (ylen)
core/includes/mail.inc:12:45 Unknown word (WINDIR)
core/includes/menu.inc:1377:19 Unknown word (mlids)
core/includes/menu.inc:1477:34 Unknown word (mlids)
core/includes/session.inc:257:63 Unknown word (pageviews)
core/includes/unicode.inc:304:18 Unknown word (myverylongurlexample)
core/includes/unicode.inc:309:34 Unknown word (myverylongurl)
core/modules/book/book.module:690:4 Unknown word (mlids)
core/modules/book/book.module:692:8 Unknown word (mlids)
core/modules/book/book.module:695:28 Unknown word (mlids)
core/modules/file/file.field.inc:1094:48 Unknown word (WSOD)
core/modules/filter/tests/filter.test:1562:57 Unknown word (litererally)

Lets fix these, so that with a single rebase all other PRs can make cspell happy.

**Comment by klonos** [comment]
OK, cspell is not complaining now. The PR only needs a review - nothing to test: https://github.com/backdrop/backdrop/pull/4537

**Comment by yorkshire-pudding** [comment]
@klonos 

> core/modules/filter/tests/filter.test:1562:57 Unknown word (litererally)

https://github.com/backdrop/backdrop/blob/a2061c2082f35ef469fa4c3fc0419ab0eeb5bf09/core/modules/filter/tests/filter.test#L1560C1-L1562

I don't fully understand whether this actual typo ("litererally" vs "literally") is crucial to the test here, but if not, wouldn't it be better to correct the typo rather than add a typo to the dictionary?

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/4537 into 1.x and 1.26.x. Thanks @klonos and @indigoxela!

**Comment by klonos** [comment]
...more words started showing up in other PRs:

Warning: Unknown word (screenreaders)
core/modules/ckeditor/js/ckeditor.admin.js:109:100 Unknown word (screenreaders)
core/modules/ckeditor/js/ckeditor.admin.js:128:100 Unknown word (screenreaders)
core/modules/link/tests/link.ui.test:347:37 Unknown word (externa)
core/modules/locale/tests/locale.test:694:[14](https://github.com/backdrop/backdrop/actions/runs/6519359244/job/17705750724?pr=4405#step:3:15) Unknown word (heure)
core/modules/locale/tests/locale.test:695:19 Unknown word (heures)
core/modules/locale/tests/locale.test:722:9 Unknown word (Ponedjeljak)
core/modules/locale/tests/locale.test:757:9 Unknown word (Ponedjeljak)
core/modules/locale/tests/locale.test:916:107 Unknown word (Svibanj)
core/modules/locale/tests/locale.test:932:87 Unknown word (Műveletek)
core/modules/locale/tests/locale.test:988:19 Unknown word (moutons)
core/modules/locale/tests/locale.test:1076:9 Unknown word (Svibanj)
core/modules/locale/tests/locale.test:11[15](https://github.com/backdrop/backdrop/actions/runs/6519359244/job/17705750724?pr=4405#step:3:16):9 Unknown word (Műveletek)
core/modules/search/tests/search.test:1114:12 Unknown word (newstr)
core/modules/search/tests/search.test:11[18](https://github.com/backdrop/backdrop/actions/runs/6519359244/job/17705750724?pr=4405#step:3:19):41 Unknown word (newstr)
core/modules/search/tests/search.test:11[19](https://github.com/backdrop/backdrop/actions/runs/6519359244/job/17705750724?pr=4405#step:3:20):14 Unknown word (newstr)
core/modules/search/tests/search.test:1119:30 Unknown word (newstr)
core/modules/search/tests/search.test:11[21](https://github.com/backdrop/backdrop/actions/runs/6519359244/job/17705750724?pr=4405#step:3:22):[25](https://github.com/backdrop/backdrop/actions/runs/6519359244/job/17705750724?pr=4405#step:3:26) Unknown word (newstr)
Warning: Unknown word (screenreaders)
Warning: Unknown word (externa)
Warning: Unknown word (heure)
Warning: Unknown word (heures)
Warning: Unknown word (Ponedjeljak)
Warning: Unknown word (Ponedjeljak)
Warning: Unknown word (Svibanj)
Warning: Unknown word (Műveletek)
Warning: Unknown word (moutons)
Warning: Unknown word (Svibanj)
Warning: Unknown word (Műveletek)
Warning: Unknown word (newstr)
Warning: Unknown word (newstr)
Warning: Unknown word (newstr)
Warning: Unknown word (newstr)
Warning: Unknown word (newstr)

Since this issue and the previous PR have been slated for 1.26.2, I'm reopening in order to file a follow-up PR to fixe the above nagging...

**Comment by klonos** [comment]
New PR here: https://github.com/backdrop/backdrop/pull/4544

**Comment by klonos** [comment]
...a few more words coming up here and there:

core/modules/search/tests/search.test:1119:14 Unknown word (newstr)
core/modules/search/tests/search.test:1119:30 Unknown word (newstr)
core/modules/search/tests/search.test:1121:25 Unknown word (newstr)
core/modules/simpletest/tests/form.test:1211:28 Unknown word (lowlevel)
core/modules/simpletest/tests/theme.test:284:16 Unknown word (parentlist)
core/modules/simpletest/tests/theme.test:298:20 Unknown word (childlist)
core/modules/simpletest/tests/theme.test:308:46 Unknown word (childlist)
core/modules/simpletest/tests/theme.test:315:27 Unknown word (parentlist)
Warning: Unknown word (newstr)
Warning: Unknown word (newstr)
Warning: Unknown word (newstr)
Warning: Unknown word (lowlevel)
Warning: Unknown word (parentlist)
Warning: Unknown word (childlist)
Warning: Unknown word (childlist)
Warning: Unknown word (parentlist)

**Comment by klonos** [comment]
...I've updated the PR:
- instead of keep adding to dictionary, I've elected to fix the "unknown" words where they are used where applicable (provided they are contained within the file/functions that were flagged).
- there's a warning re the visibility of a function not declared, but I thought best not scope creep PHPCS-related fixes into this PR

**Comment by klonos** [comment]
...here's more:

core/includes/database/mysql/schema.inc:336:6 Unknown word (fixnull)
core/includes/database/mysql/schema.inc:338:8 Unknown word (fixnull)
core/includes/database/mysql/schema.inc:352:10 Unknown word (fixnull)
core/includes/stream_wrappers.inc:785:8 Unknown word (localpath)
core/includes/stream_wrappers.inc:788:8 Unknown word (localpath)
core/includes/stream_wrappers.inc:791:21 Unknown word (localpath)
core/includes/stream_wrappers.inc:794:22 Unknown word (localpath)
core/modules/filter/filter.theme.inc:202:33 Unknown word (tiplist)
core/modules/filter/filter.theme.inc:208:18 Unknown word (tiplist)
core/modules/filter/filter.theme.inc:210:19 Unknown word (tiplist)
core/modules/node/node.module:3115:24 Unknown word (nalias)
core/modules/node/node.module:3143:24 Unknown word (nalias)
core/modules/node/node.module:3146:26 Unknown word (nalias)
core/modules/simpletest/tests/theme.test:284:16 Unknown word (parentlist)
core/modules/simpletest/tests/theme.test:298:20 Unknown word (childlist)
core/modules/simpletest/tests/theme.test:308:46 Unknown word (childlist)
core/modules/simpletest/tests/theme.test:315:27 Unknown word (parentlist)
core/modules/system/system.module:3868:4 Unknown word (zonelist)
core/modules/system/system.module:3870:13 Unknown word (zonelist)
core/modules/system/system.tar.inc:327:30 Unknown word (filelist)
core/modules/system/system.tar.inc:331:15 Unknown word (filelist)
core/modules/system/system.tar.inc:339:24 Unknown word (filelist)
core/modules/system/system.tar.inc:346:31 Unknown word (filelist)
core/modules/system/system.tar.inc:348:39 Unknown word (filelist)
core/modules/taxonomy/taxonomy.admin.inc:496:15 Unknown word (pterm)
core/modules/taxonomy/taxonomy.admin.inc:499:14 Unknown word (pterm)

PR updated again 😉

**Comment by bugfolder** [comment]
Reviewing these PRs. I observe that in many cases where compound words are created with no underscore between them, we're change core code to use underscore-delimited variables (e.g., `newstr` --> `new_string`. I see in https://github.com/backdrop/backdrop/pull/4537/files we're adding `wordsafe` to the dictionary, rather than changings its occurrences (only in `unicode.inc` and `unicode.test`) to `word_safe`, which would seem more appropriate, since even the prose expression is hyphenated ("to a word-safe return length of 20..."). So perhaps this variable should change in code, not in CSpell dictionary.

**Comment by bugfolder** [comment]
Ditto for `targetdir` --> `target_dir`, in 4 places in system.test.

**Comment by kiamlaluno** [comment]
Yes, in some cases, it makes more sense to use a better variable name. I think we should avoid adding new words in the dictionary, when possible.

**Comment by klonos** [comment]
Thanks @kiamlaluno and @bugfolder 🙏🏼  ...I have added and removed some more words to/from the dictionary, including those that you have pointed out + some duplicates that were basically pairs of the same word using a capital vs. a lower-case first letter (cspell ignores case by default anyway).

I really want us to merge this as soon as possible, in order to unblock further work on #6280, so please lets not turn this into a "lets go through all the existing words in the dictionary" exercise. We can do follow-ups for that once we have moved the dictionary to https://github.com/backdrop-ops/cspell.

Final round or reviews please?

**Comment by kiamlaluno** [comment]
I would rather not change a protected property from `$allgranularity` to `$all_granularity`, for two reasons:

- That change is not BC
- The correct property name should be `$allGranularity`, which would instead be BC, since PHP is case-insensitive

**Comment by kiamlaluno** [comment]
I would also replace *Submit handler for configing new item(s) to a view.* with * Submission handler to configure new item(s) in a view.* as to-infinitives are used to express purpose and the new items are in a view.

**Comment by quicksketch** [comment]
Overall this looks like a great set of changes, though there's still a lot of nitpicking to be done here. I'm in favor of this big renaming as a matter of overall code cleanup.

**Comment by klonos** [comment]
> - The correct property name should be `$allGranularity` (as [PHP Coding Standards](https://docs.backdropcms.org/php-standards#naming-conventions) says), which would instead be a BC change, since PHP is case-insensitive

Brilliant @kiamlaluno! 👍🏼 ...that actually works great with CSpell, which by default will flag `allgranularity` (it considers it a single word), but it recognizes `allGranularity` as a valid programming formatting of variables (camelCase). See: https://cspell.org/docs/how-it-works/#how-it-works

> The concept is simple, split camelCase and snake_case words before checking them against a list of known words.
> 
> - `camelCase` -> `camel` `case`
> - `HTMLInput` -> `html` `input`
> - `srcCode` -> `src` `code`
> - `snake_case_words` -> `snake` `case` `words`
> - `camel2snake` -> `camel` `snake` – (the 2 is ignored)
> - `function parseJson(text: string)` -> `function` `parse` `json` `text` `string`

**Comment by klonos** [comment]
> I cleaned up dozens of words starting with `a`, `b`, and `c`, ...

I am now down to letter `k` and our dictionary has currently been reduced from 541 to 417 lines! 🎉 

> ...but please lets not turn this into a "lets go through all the existing words in the dictionary" exercise. 

Well, I got sucked in it seems 😅

**Comment by klonos** [comment]
OK, I have temporarily closed https://github.com/backdrop/backdrop/pull/4544 in favor of the much smaller https://github.com/backdrop/backdrop/pull/4575 (136 vs 57 files), which still reduces our custom dictionary from 588 lines to 478.

The new PR:
- Has very minimal changes in each changed file (mostly 2-3 lines per file).
- Avoids controversial or dangerous changes that would cause pushback.
- Adds some CSpell dictionaries that will allow us to reduce the custom dictionary we are currently using: `bash`, `filetypes`, `fonts`, `publicLicenses`, `softwareTerms`, `typescript` (which includes javascript - we only had disctionaries for `css`/`html`/`php` before)
- Adds 3 whole files to the ignored paths list: `/core/includes/diff.inc`, `/core/misc/html5.js` and `/core/modules/system/system.tar.inc`. These were HUGE offenders (dozens of words removed from our custom dictionary after ignoring them), and are 3rd party files anyway.
- The majority of the changes are additions of `cspell:disable-next-line` and `cspell:disable`/`cspell:enable` pairs, that ignore unicode/non-English strings and half-words/non-words/test-words.
- Removes words in inline comments that are flagged because they are joined. For instance:
  - `Wonderwoman` -> `Wonder Woman`
  - `windows driveletter` -> `Windows drive letter`
  - `mousebutton` -> `mouse button`
  - `datastructure` -> `data structure`
  - `pageload` -> `page load`
  -  etc.
- Fixes actual typos/omissions (`dialoog` instead of `dialog`, `floo` instead of `flood`).
- Fixes words in code/URL examples, especially if they were made-up and used only once or twice in our entire codebase. For instance:
  - `http://example.com/backdropfolder/node/306` -> `http://example.com/backdrop_folder/node/306`
  - `An array of shortnames of other modules` -> `An array of machine names of other modules`
  - `allow the caller lowlevel access to the form` -> `allow the caller low-level access to the form`
  - `url('pagedoesnotexist')` -> `url('page-does-not-exist')`
  - `http://my.site.com/somepath` -> `http://my.site.com/some_path`
  - `madeup/file_test_1` -> `made_up/file_test_1`
- Splits intentionally combined words that were flagged. For example:
  - `precompute` -> `pre-compute` (and its derivatives), 
  - `ecommerce` -> `e-commerce`
  - `unioned` -> `union'ed`
- Splits unintentionally combined words, such as `more than 100 portseparators` -> `more than 100 port separators`
- Renames a single file that was used in only one test `testimage.webp` -> `test-image.webp`
- Replaces new words used in our tests with existing test words in the dictionary instead of adding more words to the dictionary. For example: `$text = "Hi there!\n-- \nHerp Derp";` -> `$text = "Hi there!\n-- \nFoo Bar";`.
- There's only 5 instances where variables are being renamed, and that is not in order to remove words from our custom dictionary, but to not add more words. These variables are either in test files, or "self-contained" within the same function

Let please fix things iteratively in order to stop the CSpell failures that are annoying contributors. Pinging our @backdrop/core-committers please.

**Comment by bugfolder** [comment]
Wow, that was  ton of work to put together. My hat is off to ya. I've reviewed the new PR, found just a few tiny tweaks. Comment when they're handled, I'll come re-review.

**Comment by klonos** [comment]
Thank you both @kiamlaluno and @bugfolder for taking the time to review, and for your suggestions. All but one are now resolved. The remaining one is the word `unioned` (lowercase - flagged by CSpell), with another 2 instances of it spelled like `UNIONed` (not flagged by CSpell) in the docblock for `???` in `core/includes/database/select.inc`.

I chose to use an apostrophe, because I'm used to `cc'ing` and then `cc'ed` or `cc'd` both being used interchangeably by most people I have been interacting with in past (I have never bothered till now to check which one is the correct/proper), but after a quick research I found these:
- https://prowritingaid.com/cc-d-or-cc-ed (claims that both are correct)
- https://www.dictionary.com/browse/cc-ed (also claims both to be correct)
- https://en.wiktionary.org/wiki/cc#English (accepts and defines both)
- https://oneminuteenglish.org/en/ccd-or-cced (concludes that both are acceptable)
- https://incorporated.zone/ccd-or-cced (concludes that both are acceptable)
- https://grammarhow.com/ccd-or-cced (concludes that both are acceptable)
- https://linguaholic.com/linguablog/cc-vs-cced-vs-copy (both)
- https://amazelaw.com/ccd-or-cced (claims that `cc'ed` is incorrect, because the `cc` abbreviation is not a verb)
- https://www.servicescape.com/blog/how-to-use-cc-as-a-verb-is-it-ccd-or-cced (also claims that `cc'ed` is incorrect)

So how do people feel about `union'd` then?

**Comment by bugfolder** [comment]
I see the situation of cc'd being different because cc is already an abbreviation, as opposed to union being an entire word. If we spelled it out, we'd say "carbon copied," not "carbon copy'd".

If "union" were a verb in everyday usage, then "unioned" would definitely be correct. It's not uncommon for technical languages to adapt common words for a specialized meaning, and per Grammarly, "Today, the following verbs-born-from-nouns are commonplace: cup, divorce, drink, dress, fool, host, intern, lure, medal, merge, model, mutter, pepper, salt, ship, sleep, strike, style, train, voice."

https://www.grammarly.com/blog/the-basics-of-verbing-nouns

Perhaps, though we should simply table this discussion. Then we can say that we've table'd it. 😉

**Comment by klonos** [comment]
>I see the situation of cc'd being different because cc is already an abbreviation, as opposed to union being an entire word.

I think that you are spot on there @bugfolder 👍🏼 ...at the end of the day:
1. This is a problem with CSpell flagging that as an unknown work. https://en.wiktionary.org/wiki/unioned for instance has it: https://en.wiktionary.org/wiki/unioned
   > **Verb**
   > **unioned**
   > 1. simple [past](https://en.wiktionary.org/wiki/Appendix:Glossary#past_tense) and [past](https://en.wiktionary.org/wiki/Appendix:Glossary#past_tense) [participle](https://en.wiktionary.org/wiki/Appendix:Glossary#participle) of [union](https://en.wiktionary.org/wiki/union#English)
2. I'd hate to delay this further for a single word. So, done! 😉

**Comment by kiamlaluno** [comment]
IMO, in the following sentence, neither *intersected* nor *unioned* makes the sentence clearer. They do not seem to add information necessary to understand the sentence.

>  Simplification: if we have both an (intersected) allowlist and a (unioned) denylist, then remove any tags from the allowlist that also exist in the denylist.

I would rather use this sentence.

> Simplification: If we have both an allow list and a deny list, remove any tag from the allow list that also exists in the deny list.

(I would actually avoid to repeat *allow list* and *deny list* in the last clause.)

**Comment by klonos** [comment]
Yes @kiamlaluno, as @bugfolder mentioned, because we went back to `unioned` (which CSpell flags as unknown), I had to re-add that word back to our custom dictionary. So once this initial PR gets merged, we still have the task to cleanup our custom dictionary from odd words that are only used once or twice in our entire codeabse, which includes the `uninoned` word. So we'll definitely cycle back to this.

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/4575 into 1.x and 1.26.x. @klonos are you inclined to continue in this issue further? Or should we open new ones?

Thanks for all these meticulous updates!

**Comment by klonos** [comment]
First of all: yay! 🎉 ❤️ 

This issue, as its title denotes, was about adding words to the custom CSpell dictionary (which ended up being adding more in-built CSpell dictionaries as well). I don't think that there are any other words that need to be added, but please allow some time to rebase the PRs that were previously failing and confirm that. If there are more words to be added, then I believe that that should be done after #6280 is merged, and it should be done by adding them to the external dictionary.

The rest of the changes (those in the temporarily closed PR over at https://github.com/backdrop/backdrop/pull/4544) are more of a clean-up of existing words rather than additions. So I think I'll file separate issues for those.

**Comment by klonos** [comment]
Quick, rather small (15 files, with no more than 5 straight-forward changes each) follow-up PR to address CSpell nags that came up after rebasing some PRs: https://github.com/backdrop/backdrop/pull/4578

**Comment by quicksketch** [comment]
Great, latest (hopefully last) PR has also been merged into 1.x and 1.26.x.

**Comment by klonos** [comment]
Lets keep this closed. I've filed #6302 for the cleanup of existing words in the dictionary, and we can use that for any odd new ones that happen to pop up.
